sandbox/cgroup/inotify/inotify_linux.go

```diff
@@ -0,0 +1,318 @@
+//go:build linux
+// +build linux
+
+// Copyright 2010 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+/*
+Package inotify implements a wrapper for the Linux inotify system.
+
+Example:
+    watcher, err := inotify.NewWatcher()
+    if err != nil {
+        log.Fatal(err)
+    }
+    err = watcher.Watch("/tmp")
+    if err != nil {
+        log.Fatal(err)
+    }
+    for {
+        select {
+        case ev := <-watcher.Event:
+            log.Println("event:", ev)
+        case err := <-watcher.Error:
+            log.Println("error:", err)
+        }
+    }
+
+*/
+package inotify
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"strings"
+	"syscall"
+	"unsafe"
+)
+
+// NewWatcher creates and returns a new inotify instance using inotify_init(2)
+func NewWatcher() (*Watcher, error) {
+	fd, errno := syscall.InotifyInit1(syscall.IN_CLOEXEC)
+	if fd == -1 {
+		return nil, os.NewSyscallError("inotify_init", errno)
+	}
+	w := &Watcher{
+		fd:      fd,
+		watches: make(map[string]*watch),
+		paths:   make(map[int]string),
+		Event:   make(chan *Event),
+		Error:   make(chan error),
+		done:    make(chan bool, 1),
+	}
+
+	go w.readEvents()
+	return w, nil
+}
+
+// Close closes an inotify watcher instance
+// It sends a message to the reader goroutine to quit and removes all watches
+// associated with the inotify instance
+func (w *Watcher) Close() error {
+	if w.isClosed {
+		return nil
+	}
+	w.isClosed = true
+
+	// Send "quit" message to the reader goroutine
+	w.done <- true
+	for path := range w.watches {
+		w.RemoveWatch(path)
+	}
```

Comment:
Mmm... so, should we move the `w.done <- true` after the loop?

Reply:
Mmm... moving the `w.done <- true` after will also result in a problem, because the loop can get stuck in the syscall.Read()...

In fact... is it already taken into account in RemoveWatch:

![imagen](https://user-images.githubusercontent.com/105285003/196976533-d9bf1c0d-7e94-46fb-9519-89f971ed839b.png)

If the descriptor is invalid, it just returns an error. At most we should set w.fd to zero and check that before calling InotifyRmWatch()... What do you think?

Reply:
Well... to zero, or to -1...

Reply:
It's a weird design: if the inotify instance is being closed, then there is no real reason to do the inotify_rm_watch calls: the watches will all be cleaned up with the inotify instance. I kind of wonder if they added those calls because they couldn't get the reader goroutine to exit without it.

I think this would be solvable, probably something like this:

1. Open the inotify instance with IN_NONBLOCK
2. Wrap the file descriptor in an `os.File` via `os.NewFile`
3. Have readEvents use the os.File.Read method rather than syscall.Read to get the next set of events.
4. Have Watcher.Close call Close on the file object. This should reliably interrupt the Read call, since it is being handled by the runtime's polling infrastructure. There'd be no need to do the rrm_watch calls.

I think we'd still need the done channel, but instead use it to unblock sends on the w.Event channel (i.e. have a select statement that tries to receive on w.done and send on w.Event).

Reply:
I think that the reason is to reuse the same code for cleaning the rest of the structures.